File: src/bridge/pdfAnalyzer.ts
// PDF Content Analyzer for Web
// Uses PDF.js to detect content type (image-only, text-based, mixed, vector-heavy)

import * as pdfjsLib from 'pdfjs-dist';
import type {
  PdfContentType,
  PdfAnalysis,
  PageContentStats,
  ReconstructionRecommendation,
} from './types';

// Ensure worker is set (may already be set by pdfParser.ts)
if (!pdfjsLib.GlobalWorkerOptions.workerSrc) {
  pdfjsLib.GlobalWorkerOptions.workerSrc = new URL(
    'pdfjs-dist/build/pdf.worker.min.mjs',
    import.meta.url
  ).toString();
}

type PDFDocument = Awaited<ReturnType<typeof pdfjsLib.getDocument>['promise']>;
type PDFPage = Awaited<ReturnType<PDFDocument['getPage']>>;

/**
 * Analyze PDF content type from Uint8Array data
 */
export async function analyzePdfContent(data: Uint8Array): Promise<PdfAnalysis> {
  const pdf = await pdfjsLib.getDocument({ data }).promise;
  const totalPages = pdf.numPages;

  if (totalPages === 0) {
    return createEmptyAnalysis();
  }

  const pageStats: PageContentStats[] = [];
  let totalText = 0;
  let totalImages = 0;
  let totalPaths = 0;
  let totalChars = 0;
  let totalImageCoverage = 0;
  let totalTextCoverage = 0;

  for (let pageNum = 1; pageNum <= totalPages; pageNum++) {
    const page = await pdf.getPage(pageNum);
    const stats = await analyzePageContent(page, pageNum - 1);

    totalText += stats.textObjects;
    totalImages += stats.imageObjects;
    totalPaths += stats.pathObjects;
    totalChars += stats.textCharCount;
    totalImageCoverage += stats.imageCoverage;
    totalTextCoverage += stats.textCoverage;

    pageStats.push(stats);
  }

  const avgImageCoverage = totalImageCoverage / totalPages;
  const avgTextCoverage = totalTextCoverage / totalPages;

  const { contentType, confidence } = classifyContent(
    totalText,
    totalImages,
    totalPaths,
    totalChars,
    avgImageCoverage,
    avgTextCoverage
  );

  const recommendation = determineRecommendation(contentType, avgImageCoverage);

  return {
    contentType,
    totalPages,
    totalTextObjects: totalText,
    totalImageObjects: totalImages,
    totalPathObjects: totalPaths,
    totalCharCount: totalChars,
    avgImageCoverage,
    avgTextCoverage,
    pageStats,
    confidence,
    recommendation,
  };
}

/**
 * Analyze a single page's content
 */
async function analyzePageContent(page: PDFPage, pageIndex: number): Promise<PageContentStats> {
  const viewport = page.getViewport({ scale: 1.0 });
  const pageArea = viewport.width * viewport.height;

  let textObjects = 0;
  let imageObjects = 0;
  let pathObjects = 0;
  let textCharCount = 0;
  let imageArea = 0;
  let textArea = 0;

  // Analyze text content
  try {
    const textContent = await page.getTextContent();
    for (const item of textContent.items) {
      if ('str' in item && item.str) {
        textObjects++;
        textCharCount += item.str.length;
        // Estimate text area from transform
        const [scaleX, , , scaleY] = item.transform;
        const fontSize = Math.abs(scaleY) || Math.abs(scaleX) || 12;
        textArea += item.width * fontSize * 1.2;
      }
    }
  } catch (e) {
    console.warn('Failed to get text content:', e);
  }

  // Analyze operator list for images and paths
  try {
    const operatorList = await page.getOperatorList();
    const OPS = pdfjsLib.OPS;
    const ctmStack: number[][] = [[1, 0, 0, 1, 0, 0]];
    let currentCTM = [1, 0, 0, 1, 0, 0];

    for (let i = 0; i < operatorList.fnArray.length; i++) {
      const fn = operatorList.fnArray[i];
      const args = operatorList.argsArray[i];

      if (fn === OPS.save) {
        ctmStack.push([...currentCTM]);
      } else if (fn === OPS.restore) {
        currentCTM = ctmStack.pop() || [1, 0, 0, 1, 0, 0];
      } else if (fn === OPS.transform) {
        currentCTM = multiplyMatrices(currentCTM, args as number[]);
      } else if (fn === OPS.paintImageXObject || fn === OPS.paintXObject) {
        imageObjects++;
        // Estimate image area from CTM
        const [a, , , d] = currentCTM;
        imageArea += Math.abs(a * d);
      } else if (
        fn === OPS.moveTo ||
        fn === OPS.lineTo ||
        fn === OPS.curveTo ||
        fn === OPS.rectangle ||
        fn === OPS.stroke ||
        fn === OPS.fill
      ) {
        pathObjects++;
      }
    }
  } catch (e) {
    console.warn('Failed to analyze operator list:', e);
  }

  return {
    pageIndex,
    textObjects,
    imageObjects,
    pathObjects,
    textCharCount,
    imageCoverage: Math.min(imageArea / pageArea, 1.0),
    textCoverage: Math.min(textArea / pageArea, 1.0),
  };
}

/**
 * Classify PDF content type based on statistics
 */
function classifyContent(
  textObjects: number,
  imageObjects: number,
  pathObjects: number,
  charCount: number,
  avgImageCoverage: number,
  _avgTextCoverage: number
): { contentType: PdfContentType; confidence: number } {
  // Image-only: high image coverage, minimal text
  if (avgImageCoverage > 0.7 && charCount < 50 && textObjects < 5) {
    return { contentType: 'image-only', confidence: 0.95 };
  }

  // Text-based: significant text, low image coverage
  if (charCount > 100 && avgImageCoverage < 0.2 && textObjects > imageObjects * 2) {
    return { contentType: 'text-based', confidence: 0.9 };
  }

  // Vector-heavy: many path objects, low text/image
  if (pathObjects > (textObjects + imageObjects) * 2 && pathObjects > 50) {
    return { contentType: 'vector-heavy', confidence: 0.85 };
  }

  // Mixed: significant both text and images
  if (charCount > 50 && imageObjects > 0 && avgImageCoverage > 0.1) {
    return { contentType: 'mixed', confidence: 0.8 };
  }

  // Default to text-based if there's any text
  if (charCount > 0 || textObjects > 0) {
    return { contentType: 'text-based', confidence: 0.7 };
  }

  return { contentType: 'empty', confidence: 0.5 };
}

/**
 * Determine reconstruction recommendation
 */
function determineRecommendation(
  contentType: PdfContentType,
  avgImageCoverage: number
): ReconstructionRecommendation {
  switch (contentType) {
    case 'image-only':
      return 'ocr-required';
    case 'mixed':
      return avgImageCoverage > 0.5 ? 'ocr-verification' : 'none';
    case 'vector-heavy':
      return 'vector-conversion';
    default:
      return 'none';
  }
}

/**
 * Create empty analysis result
 */
function createEmptyAnalysis(): PdfAnalysis {
  return {
    contentType: 'empty',
    totalPages: 0,
    totalTextObjects: 0,
    totalImageObjects: 0,
    totalPathObjects: 0,
    totalCharCount: 0,
    avgImageCoverage: 0,
    avgTextCoverage: 0,
    pageStats: [],
    confidence: 1.0,
    recommendation: 'none',
  };
}

/**
 * Multiply two transformation matrices
 */
function multiplyMatrices(m1: number[], m2: number[]): number[] {
  return [
    m1[0] * m2[0] + m1[2] * m2[1],
    m1[1] * m2[0] + m1[3] * m2[1],
    m1[0] * m2[2] + m1[2] * m2[3],
    m1[1] * m2[2] + m1[3] * m2[3],
    m1[0] * m2[4] + m1[2] * m2[5] + m1[4],
    m1[1] * m2[4] + m1[3] * m2[5] + m1[5],
  ];
}

/**
 * Get human-readable description of content type
 */
export function getContentTypeDescription(contentType: PdfContentType): string {
  switch (contentType) {
    case 'image-only':
      return 'Scanned Document (Image-Only)';
    case 'text-based':
      return 'Native Text PDF';
    case 'mixed':
      return 'Mixed Content (Text + Images)';
    case 'vector-heavy':
      return 'Vector Graphics Heavy';
    case 'empty':
      return 'Empty Document';
  }
}

/**
 * Get human-readable recommendation text
 */
export function getRecommendationText(recommendation: ReconstructionRecommendation): string {
  switch (recommendation) {
    case 'ocr-required':
      return 'OCR recommended to extract text from images';
    case 'ocr-verification':
      return 'OCR verification available to improve accuracy';
    case 'vector-conversion':
      return 'Vector elements can be converted to editable shapes';
    case 'none':
      return 'No reconstruction needed';
  }
}

File: src-tauri/src/pdf_analyzer.rs
//! PDF Content Analyzer Module
//!
//! Detects PDF content type (image-only, text-based, mixed, vector-heavy)
//! and provides reconstruction strategies.

use pdfium_render::prelude::*;
use serde::{Deserialize, Serialize};

/// PDF content type classification
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum PdfContentType {
    /// Scanned document - only images, no extractable text
    ImageOnly,
    /// Native text PDF - primarily text with optional images
    TextBased,
    /// Mixed content - significant text and images
    Mixed,
    /// Vector graphics heavy - diagrams, charts, illustrations
    VectorHeavy,
    /// Empty or unreadable
    Empty,
}

/// Content statistics for a single page
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageContentStats {
    pub page_index: usize,
    pub text_objects: usize,
    pub image_objects: usize,
    pub path_objects: usize,
    pub text_char_count: usize,
    /// Ratio of image area to page area (0.0 - 1.0)
    pub image_coverage: f32,
    /// Ratio of text area to page area (0.0 - 1.0)  
    pub text_coverage: f32,
}

/// Complete PDF analysis result
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PdfAnalysis {
    pub content_type: PdfContentType,
    pub total_pages: usize,
    pub total_text_objects: usize,
    pub total_image_objects: usize,
    pub total_path_objects: usize,
    pub total_char_count: usize,
    pub avg_image_coverage: f32,
    pub avg_text_coverage: f32,
    pub page_stats: Vec<PageContentStats>,
    /// Confidence score for the classification (0.0 - 1.0)
    pub confidence: f32,
    /// Recommended action based on analysis
    pub recommendation: ReconstructionRecommendation,
}

/// Reconstruction recommendation
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum ReconstructionRecommendation {
    /// No reconstruction needed - text is extractable
    None,
    /// Run OCR on image-only pages
    OcrRequired,
    /// Enhance text extraction with OCR verification
    OcrVerification,
    /// Vector to editable conversion available
    VectorConversion,
}

/// Analyze a PDF file and return content classification
pub fn analyze_pdf(file_path: &str) -> Result<PdfAnalysis, String> {
    let pdfium = Pdfium::default();
    let document = pdfium
        .load_pdf_from_file(file_path, None)
        .map_err(|e| format!("Failed to load PDF: {}", e))?;

    let total_pages = document.pages().len() as usize;
    if total_pages == 0 {
        return Ok(PdfAnalysis {
            content_type: PdfContentType::Empty,
            total_pages: 0,
            total_text_objects: 0,
            total_image_objects: 0,
            total_path_objects: 0,
            total_char_count: 0,
            avg_image_coverage: 0.0,
            avg_text_coverage: 0.0,
            page_stats: vec![],
            confidence: 1.0,
            recommendation: ReconstructionRecommendation::None,
        });
    }

    let mut page_stats = Vec::with_capacity(total_pages);
    let mut total_text = 0usize;
    let mut total_images = 0usize;
    let mut total_paths = 0usize;
    let mut total_chars = 0usize;
    let mut total_image_coverage = 0.0f32;
    let mut total_text_coverage = 0.0f32;

    for page_idx in 0..total_pages {
        let page = document
            .pages()
            .get(page_idx as u16)
            .map_err(|e| format!("Failed to get page {}: {}", page_idx, e))?;

        let stats = analyze_page(&page, page_idx);
        
        total_text += stats.text_objects;
        total_images += stats.image_objects;
        total_paths += stats.path_objects;
        total_chars += stats.text_char_count;
        total_image_coverage += stats.image_coverage;
        total_text_coverage += stats.text_coverage;
        
        page_stats.push(stats);
    }

    let avg_image_coverage = total_image_coverage / total_pages as f32;
    let avg_text_coverage = total_text_coverage / total_pages as f32;

    // Classify content type
    let (content_type, confidence) = classify_content(
        total_text,
        total_images,
        total_paths,
        total_chars,
        avg_image_coverage,
        avg_text_coverage,
    );

    let recommendation = determine_recommendation(&content_type, avg_image_coverage);

    Ok(PdfAnalysis {
        content_type,
        total_pages,
        total_text_objects: total_text,
        total_image_objects: total_images,
        total_path_objects: total_paths,
        total_char_count: total_chars,
        avg_image_coverage,
        avg_text_coverage,
        page_stats,
        confidence,
        recommendation,
    })
}

/// Analyze a single page
fn analyze_page(page: &PdfPage, page_index: usize) -> PageContentStats {
    let page_width = page.width().value as f32;
    let page_height = page.height().value as f32;
    let page_area = page_width * page_height;

    let mut text_objects = 0usize;
    let mut image_objects = 0usize;
    let mut path_objects = 0usize;
    let mut text_char_count = 0usize;
    let mut image_area = 0.0f32;
    let mut text_area = 0.0f32;

    for object in page.objects().iter() {
        match object.object_type() {
            PdfPageObjectType::Text => {
                text_objects += 1;
                if let Some(text_obj) = object.as_text_object() {
                    text_char_count += text_obj.text().chars().count();
                    if let Ok(bounds) = text_obj.bounds() {
                        let w = (bounds.right().value - bounds.left().value) as f32;
                        let h = (bounds.top().value - bounds.bottom().value) as f32;
                        text_area += w * h;
                    }
                }
            }
            PdfPageObjectType::Image => {
                image_objects += 1;
                if let Some(image_obj) = object.as_image_object() {
                    if let Ok(bounds) = image_obj.bounds() {
                        let w = (bounds.right().value - bounds.left().value) as f32;
                        let h = (bounds.top().value - bounds.bottom().value) as f32;
                        image_area += w * h;
                    }
                }
            }
            PdfPageObjectType::Path => {
                path_objects += 1;
            }
            _ => {}
        }
    }

    PageContentStats {
        page_index,
        text_objects,
        image_objects,
        path_objects,
        text_char_count,
        image_coverage: (image_area / page_area).min(1.0),
        text_coverage: (text_area / page_area).min(1.0),
    }
}

/// Classify PDF content type based on statistics
fn classify_content(
    text_objects: usize,
    image_objects: usize,
    path_objects: usize,
    char_count: usize,
    avg_image_coverage: f32,
    _avg_text_coverage: f32,
) -> (PdfContentType, f32) {
    // Image-only: high image coverage, minimal text
    if avg_image_coverage > 0.7 && char_count < 50 && text_objects < 5 {
        return (PdfContentType::ImageOnly, 0.95);
    }

    // Text-based: significant text, low image coverage
    if char_count > 100 && avg_image_coverage < 0.2 && text_objects > image_objects * 2 {
        return (PdfContentType::TextBased, 0.9);
    }

    // Vector-heavy: many path objects, low text/image
    if path_objects > (text_objects + image_objects) * 2 && path_objects > 50 {
        return (PdfContentType::VectorHeavy, 0.85);
    }

    // Mixed: significant both text and images
    if char_count > 50 && image_objects > 0 && avg_image_coverage > 0.1 {
        return (PdfContentType::Mixed, 0.8);
    }

    // Default to text-based if there's any text
    if char_count > 0 || text_objects > 0 {
        return (PdfContentType::TextBased, 0.7);
    }

    // Empty or unclassifiable
    (PdfContentType::Empty, 0.5)
}

/// Determine reconstruction recommendation
fn determine_recommendation(
    content_type: &PdfContentType,
    avg_image_coverage: f32,
) -> ReconstructionRecommendation {
    match content_type {
        PdfContentType::ImageOnly => ReconstructionRecommendation::OcrRequired,
        PdfContentType::Mixed if avg_image_coverage > 0.5 => {
            ReconstructionRecommendation::OcrVerification
        }
        PdfContentType::VectorHeavy => ReconstructionRecommendation::VectorConversion,
        _ => ReconstructionRecommendation::None,
    }
}

/// Tauri command to analyze PDF
#[tauri::command]
pub async fn analyze_pdf_content(file_path: String) -> Result<PdfAnalysis, String> {
    analyze_pdf(&file_path)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_classify_image_only() {
        let (content_type, _) = classify_content(2, 5, 0, 10, 0.85, 0.01);
        assert_eq!(content_type, PdfContentType::ImageOnly);
    }

    #[test]
    fn test_classify_text_based() {
        let (content_type, _) = classify_content(100, 2, 10, 5000, 0.05, 0.6);
        assert_eq!(content_type, PdfContentType::TextBased);
    }

    #[test]
    fn test_classify_mixed() {
        let (content_type, _) = classify_content(50, 10, 20, 500, 0.3, 0.4);
        assert_eq!(content_type, PdfContentType::Mixed);
    }

    #[test]
    fn test_recommendation_ocr() {
        let rec = determine_recommendation(&PdfContentType::ImageOnly, 0.9);
        assert_eq!(rec, ReconstructionRecommendation::OcrRequired);
    }
}
